### PR TITLE
CART-822 test: Add tools to check/fix D_FREE usage.

### PR DESCRIPTION
**Pull request by ashleypittman**

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -355,7 +355,7 @@ fetch_entry(daos_handle_t oh, daos_handle_t th, const char *name,
 		*exists = true;
 
 out:
-	if (fetch_sym)
+	if (value)
```
Comment (jolivier23) [inline src/client/dfs/dfs.c:358]: do you need the check?   D_FREE handles NULL

File: ci/check_d_macro_calls.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        """Move the code on a line left one tabstop."""
+        self._set_text(self.text[1:])
+
+    def remove_free_ptr(self):
+        """Replace use of D_FREE_PTR"""
+        if 'D_FREE_PTR' in self.code:
+            self._set_text(self._get_text().replace('D_FREE_PTR', 'D_FREE'))
+
+prev_file = None
+def show_patch(lines):
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply ci/check_d_macro_calls.py:168]: missed one

File: ci/check_d_macro_calls.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    # D_FREE_PTR
+    for line in lines:
+        line.remove_free_ptr()
+
+    # Strip out and pre and post-amble, patch does not like it if these
+    # are uneven length, and as patches are not long-lived there is
+    # little benefit in keeping them.
+    start = len(lines)
+    end = 0
+    for idx in range(len(lines)):
```
[the remark this answers is not in the document]
Comment (jolivier23) [reply ci/check_d_macro_calls.py:196]: silence it?

File: src/vos/vos_io.c [hunk 4]
```diff
@@ -195,7 +195,8 @@ vos_ioc_destroy(struct vos_io_context *ioc, bool evict)
 	vos_ilog_fetch_finish(&ioc->ic_dkey_info);
 	vos_ilog_fetch_finish(&ioc->ic_akey_info);
 	vos_cont_decref(ioc->ic_cont);
-	vos_ts_set_free(ioc->ic_ts_set);
+	if (ioc->ic_ts_set)
```
Comment (jolivier23) [inline src/vos/vos_io.c:198]: is it needed?   We've removed such checks in many places

**Comment by ashleypittman** [comment]
> so does this replace it or just check it?

This just adds checks.

It extends the log-checking code to find locations where we're calling D_FREE(NULL), and sort locations by frequency, the idea being that we can remove all the checks, then either rework code or put back in checks for locations which do this a lot (the group and register code both have locations where they call this 100s of times).
It also adds a script which can modify the code in-place to remove locations where this is done, I'm hoping to land this, then in another PR I'll fix the locations, and enable the check, but we probably want a few iterations to add a whitelist for the group code first.
There are some cases like in the dfuse code where we can silence otherwise noisy calls by changing the D_GOTO line.

File: src/client/dfs/dfs.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfs/dfs.c:358]: This is one location that's reporting a lot, in fact there's something odd in the handling of value in this function which I'm going to fix-up elsewhere.  This line is about 20% of the DFS/dfuse logging on metadata lookups.

File: src/vos/vos_io.c (hunk 4 above)
Comment (ashleypittman) [reply src/vos/vos_io.c:198]: This is another case where we're seeing 100s of log lines.  They were initially all reported in the header file but as it's a one-line function I made it a macro and they all originate here.

File: ci/check_d_macro_calls.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ashleypittman) [reply ci/check_d_macro_calls.py:196]: I will, it's a new one on me though, I'll have to look up how to do it.

**Review by jolivier23 — APPROVED** [review]

**Pull request closed without merging**